dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java

```diff
@@ -0,0 +1,249 @@
+            excluded.sort(Comparator.naturalOrder());
+
+            if (oldIncluded.equals(included) && oldExcluded.equals(excluded) && newConfigMode == oldConfigMode) {
+                Tr.event(this, tc, "After update there is no need to rebuild the openAPI document");
+            } else {
+                Tr.event(this, tc, "Clearing out the outdated openAPI document");
+
+                openAPIAppConfigListeners.sort((OpenAPIAppConfigListener o1, OpenAPIAppConfigListener o2) -> Integer.compare(o1.getConfigListenerPriority(),
+                                                                                                                             o2.getConfigListenerPriority()));
+
+                for (OpenAPIAppConfigListener listener : openAPIAppConfigListeners) {
+                    listener.processConfigUpdate();
+                }
+            }
+        }
+    }
+
+    @Deactivate
+    protected void deactivate() {
+        if (ProductInfo.getBetaEdition()) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(this, tc, "Deactivating OpenAPIAppConfigProviderImpl");
+            }
+            includedApps.clear();
+            excludedApps.clear();
+            includedModules.clear();
+            excludedModules.clear();
+        }
+    }
+
+    private void betaFenceCheck() throws UnsupportedOperationException {
+        // Not running beta edition, throw exception
+        if (!ProductInfo.getBetaEdition()) {
+            throw new UnsupportedOperationException("This method is beta and is not available.");
+        } else {
+            // Running beta exception, issue message if we haven't already issued one for this class
+            if (!issuedBetaMessage) {
+                Tr.info(tc, "BETA: A beta method has been invoked for the class " + this.getClass().getName() + " for the first time.");
+                issuedBetaMessage = !issuedBetaMessage;
+            }
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Deprecated
+    public Optional<List<String>> getIncludedAppsAndModules() {
+        betaFenceCheck();
+        List<String> combinedList = new ArrayList<String>(includedApps);
+        combinedList.addAll(includedModules);
+        return combinedList.isEmpty() ? Optional.<List<String>> empty() : Optional.of(Collections.unmodifiableList(combinedList));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Deprecated
+    public Optional<List<String>> getExcludedAppsAndModules() {
+        betaFenceCheck();
+        List<String> combinedList = new ArrayList<String>(excludedApps);
+        combinedList.addAll(excludedModules);
+        return combinedList.isEmpty() ? Optional.<List<String>> empty() : Optional.of(Collections.unmodifiableList(combinedList));
+    }
+
+    @Override
+    @Deprecated
+    public Optional<ConfigMode> getConfigMode() {
+        betaFenceCheck();
+        return configMode;
+    }
```

Comment:
We have a thread safety issue here (and in other places in the class).

When the config is modified, fields `includedApps` and `includedModules` get updated. If they're modified while we're copying their values (in the `ArrayList` constructor or the `addAll` method), we could get an exception.

Secondly, because there are three separate methods here, there's no way for the caller to get a consistent view of the config. The values could be updated between a call to the first method and a call to the second.

As you suggested, we could fix this with an inner class which holds a snapshot of the config and a method which returns that.